File: app/src/main/java/com/lostincontext/playlists/PlaylistsAdapter.java
package com.lostincontext.playlists;


import android.support.annotation.IdRes;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.lostincontext.R;
import com.lostincontext.commons.list.EmptyListCallback;
import com.lostincontext.commons.list.StatefulAdapter;
import com.lostincontext.commons.list.ViewHolder;
import com.lostincontext.data.playlist.Playlist;

import java.util.List;

import static java.util.Collections.emptyList;

public class PlaylistsAdapter extends StatefulAdapter {

    private List<Playlist> playlists = emptyList();

    private PlaylistViewHolder.Callback itemCallback;
    private EmptyListCallback emptyListCallback;

    public PlaylistsAdapter(PlaylistViewHolder.Callback itemCallback,
                            EmptyListCallback emptyListCallback) {
        super(ContentState.LOADING);
        this.itemCallback = itemCallback;
        this.emptyListCallback = emptyListCallback;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, @IdRes int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        switch (viewType) {
            case R.id.view_type_standard:
                return PlaylistViewHolder.create(layoutInflater, parent, itemCallback);

            case R.id.view_type_loading:
                return buildLoadingViewHolder(layoutInflater, parent);

            case R.id.view_type_error:
                return buildErrorViewHolder(layoutInflater, parent, emptyListCallback);

            case R.id.view_type_empty:
                return buildEmptyViewHolder(layoutInflater, parent);

            default:
                throw new IllegalStateException("the adapter is in an invalid state");
        }
    }

    @Override public void onBindViewHolder(ViewHolder holder, int position, List<Object> payloads) {
        switch (holder.getItemViewType()) {
            case R.id.view_type_standard:
                PlaylistViewHolder viewHolder = (PlaylistViewHolder) holder;
                viewHolder.bindTo(playlists.get(position));
                break;


            case R.id.view_type_loading:
            case R.id.view_type_error: //todo
            case R.id.view_type_empty:
                // nothing to do for these
                break;

            default:
                throw new IllegalStateException("the adapter is in an invalid state");
        }

    }


    @Override public int getContentItemsCount() {
        return playlists.size();
    }


    public void setPlaylists(List<Playlist> playlists) {
        this.playlists = playlists;
        if (playlists.size() == 0) setCurrentState(ContentState.EMPTY);
        else setCurrentState(ContentState.CONTENT);
    }
}

File: app/src/main/java/com/lostincontext/playlists/PlaylistsAdapter.kt
package com.lostincontext.playlists


import android.support.annotation.IdRes
import android.support.v7.util.DiffUtil
import android.view.LayoutInflater
import android.view.ViewGroup
import com.lostincontext.R
import com.lostincontext.commons.list.*
import com.lostincontext.commons.list.StatefulAdapter.ContentState.*
import com.lostincontext.data.playlist.Playlist

class PlaylistsAdapter(private val itemCallback: PlaylistViewHolder.Callback,
                       private val emptyListCallback: EmptyListCallback,
                       private val infiniteAdapterCallback: InfiniteAdapterCallback)
: StatefulAdapter(LOADING) {

    private var playlists: List<Playlist> = emptyList()
    private var totalNumberOfItems: Int = 0
    private var isLoadingMoreItems = false

    override fun onCreateViewHolder(parent: ViewGroup, @IdRes viewType: Int): ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        when (viewType) {
            R.id.view_type_standard -> return PlaylistViewHolder.create(layoutInflater,
                                                                        parent,
                                                                        itemCallback)

            R.id.view_type_loading -> return buildLoadingViewHolder(layoutInflater,
                                                                    parent)

            R.id.view_type_error -> return buildErrorViewHolder(layoutInflater,
                                                                parent,
                                                                emptyListCallback)

            R.id.view_type_empty -> return buildEmptyViewHolder(layoutInflater,
                                                                parent)

            R.id.view_type_load_more -> return createLoadMoreViewHolder(layoutInflater, parent)

            else -> throw IllegalStateException("the adapter is in an invalid state")
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
        when (holder.itemViewType) {
            R.id.view_type_standard -> {
                val viewHolder = holder as PlaylistViewHolder
                viewHolder.bindTo(playlists[position])

                if (position > playlists.size - LOAD_MORE_THRESHOLD
                        && canLoadMore()
                        && !isLoadingMoreItems) {
                    isLoadingMoreItems = true
                    infiniteAdapterCallback.onNeedMoreItems()
                }
            }


            R.id.view_type_loading,
            R.id.view_type_load_more,
            R.id.view_type_error, //todo
            R.id.view_type_empty -> {
            }

            else -> throw IllegalStateException("the adapter is in an invalid state")
        }

    }

    override fun getContentItemsCount(): Int = playlists.size + if (canLoadMore()) 1 else 0

    override fun getContentItemViewType(position: Int): Int {
        if (canLoadMore() && position == playlists.size) return R.id.view_type_load_more
        return R.id.view_type_standard
    }

    fun setPlaylists(playlists: List<Playlist>, total: Int) {
        this.totalNumberOfItems = total
        this.playlists = playlists
        if (playlists.size == 0) setCurrentState(EMPTY)
        else setCurrentState(CONTENT, false)
    }

    fun canLoadMore(): Boolean = totalNumberOfItems > playlists.size


    /**
     * swap the current list for a new one and dispatches the notify* calls
     */
    fun swap(newPlaylists: List<Playlist>) {

        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                if (oldItemPosition >= playlists.size || newItemPosition >= newPlaylists.size) {
                    return false
                }
                return playlists[oldItemPosition] == newPlaylists[newItemPosition]
            }

            override fun getOldListSize(): Int = itemCount


            override fun getNewListSize(): Int = newPlaylists.size + if (canLoadMore()) 1 else 0

            override fun areContentsTheSame(oldItemPosition: Int,
                                            newItemPosition: Int): Boolean = true

        }, false)
        playlists = newPlaylists
        diffResult.dispatchUpdatesTo(this)
        isLoadingMoreItems = false
    }


    companion object {
        const val LOAD_MORE_THRESHOLD = 5
    }


}
